engine/execution/ingestion/stop_at_height.go

```diff
@@ -0,0 +1,167 @@
+package ingestion
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/engine/execution/state"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// StopControl is a specialized component used by ingestion.Engine to encapsulate
+// control of pausing/stopping blocks execution.
+type StopControl struct {
+	sync.RWMutex
+	height             uint64
+	crash              bool
+	commenced          bool // if stopping process has started. We disallow any changes then.
+	set                bool // whether stop at height has been set at all, it its envisioned most of the time it won't
+	paused             bool // if execution is paused
+	stopAfterExecuting flow.Identifier
+	log                zerolog.Logger
+}
+
+// NewStopControl creates new empty NewStopControl
+func NewStopControl(log zerolog.Logger, paused bool) *StopControl {
+	return &StopControl{
+		paused:    paused,
+		commenced: paused,
+		log:       log,
+	}
+}
+
+// IsPaused returns true is block execution has been paused
+func (s *StopControl) IsPaused() bool {
+	s.RLock()
+	defer s.RUnlock()
+	return s.paused
+}
+
+// SetStopHeight sets new stop height and crash mode, and return old values:
+//   - set, whether values were previously set
+//   - height
+//   - crash
+//
+// Returns error is the stopping process has already commenced, new values will be rejected.
+func (s *StopControl) SetStopHeight(height uint64, crash bool) (bool, uint64, bool, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	oldSet := s.set
+	oldHeight := s.height
+	oldCrash := s.crash
+
+	if s.commenced {
+		return oldSet, oldHeight, oldCrash, fmt.Errorf("cannot update stop height, stopping already in progress for height %d with crash=%t", oldHeight, oldCrash)
+	}
+
+	s.set = true
+	s.height = height
+	s.crash = crash
+	s.stopAfterExecuting = flow.ZeroID
+
+	return oldSet, oldHeight, oldCrash, nil
+}
+
+// GetStopHeight returns:
+//   - set, whether values are set
+//   - height
+//   - crash
+func (s *StopControl) GetStopHeight() (bool, uint64, bool) {
+	s.RLock()
+	defer s.RUnlock()
+
+	return s.set, s.height, s.crash
+}
+
+// BlockProcessable should be called when new block is processable.
+// It returns boolean indicating if processing should skip this block.
+func (s *StopControl) BlockProcessable(b *flow.Header) bool {
+
+	s.Lock()
+	defer s.Unlock()
+
+	if !s.set {
+		return false
+	}
+
+	if s.paused {
+		return true
+	}
+
+	// skips blocks at or above requested stop height
+	if b.Height >= s.height {
+		s.log.Warn().Msgf("Skipping execution of %s at height %d because stop has been requested at height %d", b.ID(), b.Height, s.height)
+		s.commenced = true
```

Comment:
This leads me to think, what ensures that before setting the stop height, there isn't a block above the stop height has been executed?

If block 101 has been executed, and I set stop executing at height 100, then when switching version after restart, block 101 will be skipped as it's already executed by old version. 

We have a Highest executed height saved in database, should we use that?

Reply:
Thats a good question. 
In the version contract we explicitly disallow changes to blocks within 1000 blocks from current one.
Maybe that would work here as well - so disallow setting within X blocks of latest executed (I think there can be some edge cases if we disallow on last executed height directly, forks, etc)

Reply:
I have added height tracking now, which should prevent this problem.